src/include/mpir_handlemem.h

```diff
@@ -262,14 +263,18 @@ static inline void *MPIR_Handle_obj_alloc_unsafe(MPIR_Object_alloc_t * objmem,
             if (ptr) {
                 objmem->avail = ptr->next;
             }
+            /* MPIR_Add_finalize() accesses a global data structure, so some
+             * MPIR_Handle_obj_alloc_xxx() functions do not want to call it. */
+            if (!skip_finalize) {
 #ifdef MPICH_DEBUG_HANDLEALLOC
-            /* The priority of these callbacks must be greater than
-             * the priority of the callback that frees the objmem direct and
-             * indirect storage. */
-            MPIR_Add_finalize(MPIR_check_handles_on_finalize, objmem,
-                              MPIR_FINALIZE_CALLBACK_HANDLE_CHECK_PRIO);
+                /* The priority of these callbacks must be greater than
+                 * the priority of the callback that frees the objmem direct and
+                 * indirect storage. */
+                MPIR_Add_finalize(MPIR_check_handles_on_finalize, objmem,
+                                  MPIR_FINALIZE_CALLBACK_HANDLE_CHECK_PRIO);
 #endif
-            MPIR_Add_finalize(MPIR_Handle_finalize, objmem, 0);
+                MPIR_Add_finalize(MPIR_Handle_finalize, objmem, 0);
```

Comment:
This is an issue. How do we clean up the dymnically allocated `handle_mem` slabs?

Reply:
I think we should retain this `MPIR_Handle_finalize`, which will free the allocated slabs. That doesn't prevent new info being created after the finalize.

Reply:
We may need a separate `MPIR_Info_handle_finalize` if `MPIR_Handle_finalize` always free the memory without checking whether it's still in use.

Reply:
I didn't really understand this mechanism. Yes, you're right.

I believe the performance is not important.  I came up with a few possible implementations, but all need nontrivial modification.

1. When `MPIR_Handle_obj_free_noinit()` is called, we can always count alive objects. If all handles are returned, we can manually call MPIR_Handle_finalize().

2. I can create a totally different set of `MPIR_Handle_obj_xxx()` that dynamically allocates and frees memory for each MPI_Info object.

3. The size of MPI_Info becomes sizeof(void *).  Then we can assign allocated memory.

4. Retain `MPIR_Handle_finalize()` (as you suggested).  This causes SEGV if you want to refer to an INFO object that has been allocated before `MPI_Finalize()`.  With a debugging option, it may cause "memory leak" warning in `MPIR_check_handles_on_finalize()` if not all INFO objects are freed before `MPI_Finalize()`.  In addition, it causes real memory leak if `MPI_Info_create_env()` is called many times after `MPI_Finalize()`.

How about 1? If 4 (you suggested) is fine so far, I will go for it.

Reply:
Based on the discussion, I implemented it properly since MPI-4 requires the implementation to allow any MPI_Info functions before `MPI_Initialize()` and after `MPI_Finalize()`. The current implementation is based on the idea 1.